all: add Go 1.24 support

aykevl opened the pull request:
Not entirely finished, but let's see how far we get with just a few adjustments.

aykevl committed:
all: add Go 1.24 support

deadprogram commented:
Looks like the official image for Go 1.24 is not yet available on https://hub.docker.com/_/golang/tags

aykevl commented:
Yeah you're right, I hope they add it in a day or so.

aykevl commented:
Perhaps we can use the 1.24rc3 one for now?

deadprogram commented:
> Perhaps we can use the 1.24rc3 one for now?

A very good idea.

joonas commented:
FWIW, it looks like the official [`1.24.0` tag (and it's variants)](https://hub.docker.com/_/golang/tags?name=1.24.0) landed about 2 hours ago 🙂

aykevl commented:
I'm investigating a weird CGo problem now.

aykevl commented:
The CGo regression is a result of this commit: https://github.com/golang/go/commit/5bd442aba31c1720bb333dafbb65cf977091d292
Need to think about the right way to proceed. Is it our bug, or a bug in Go?

aykevl commented:
Filed an upstream bug: https://github.com/golang/go/issues/71777

aykevl committed:
cgo: mangle identifier names

This mangles CGo identifier names to something like "_Cgo_foo" instead
of using literal identifiers like "C.foo". This works around
https://github.com/golang/go/issues/71777.

I don't like this solution, but I hope we'll find a better solution in
the future. In that case we can revert this commit.

aykevl commented:
Added a workaround for the CGo issue. This will regress our nice CGo error messages, but I don't see a good alternative.

Previous error message:

```
# command-line-arguments
tmp/test.go:7:6: undefined: C.FOO
```

Error message with the workaround:
```
# command-line-arguments
tmp/test.go:7:6: undefined: _Cgo_FOO
```

aykevl committed:
machine: remove bytes package dependency in flash code

This also moves flash padding code to a single place, since it was
copied 5 times.

This change is necessary in Go 1.24 to avoid an import cycle.

aykevl committed:
machine/usb/descriptor: avoid bytes package

We can't use the bytes package in Go 1.24 since it would result in an
import cycle. Therefore, use bytealg.Index instead.

(I'm not sure this code is correct - just searching for a range of bytes
seems brittle. But at least this commit shouldn't change the code).

aykevl commented on GNUmakefile:
```diff
@@ -355,13 +355,6 @@ TEST_PACKAGES_FAST = \
 	unique \
 	$(nil)
 
-# Assume this will go away before Go2, so only check minor version.
-ifeq ($(filter $(shell $(GO) env GOVERSION | cut -f 2 -d.), 16 17 18), )
-TEST_PACKAGES_FAST += crypto/internal/nistec/fiat
-else
-TEST_PACKAGES_FAST += crypto/elliptic/internal/fiat
-endif
-
```
I removed these, since they're an internal package and the name keeps changing. In Go 1.24, the name seems to have changed to crypto/internal/fips140/nistec/fiat.

ydnar commented:
```console
# text/template
/usr/local/go/src/text/template/exec.go:405:22: val.Seq undefined (type reflect.Value has no field or method Seq)
/usr/local/go/src/text/template/exec.go:454:17: val.Type().CanSeq undefined (type reflect.Type has no field or method CanSeq)
/usr/local/go/src/text/template/exec.go:460:23: val.Seq undefined (type reflect.Value has no field or method Seq)
/usr/local/go/src/text/template/exec.go:471:17: val.Type().CanSeq2 undefined (type reflect.Type has no field or method CanSeq2)
/usr/local/go/src/text/template/exec.go:473:26: val.Seq2 undefined (type reflect.Value has no field or method Seq2)
```

I can peek at this.

aykevl commented:
Fixed two issues (I hope), now one that looks scary:

```
~/src/tinygo/tinygo$ tinygo test -target=wasip1 compress/lzw
Error: failed to run main module `/tmp/tinygo4120297224/main.wasmopt`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x13dd0 - runtime.abort
                           at /home/ayke/src/tinygo/tinygo/src/runtime/runtime_tinygowasm.go:78:6              - runtime.runtimePanicAt
                           at /home/ayke/src/tinygo/tinygo/src/runtime/panic.go:103:7
           1: 0x56c7 - runtime.nilPanic
                           at /home/ayke/src/tinygo/tinygo/src/runtime/panic.go:180:16
           2: 0x1cbd9 - (*flag.FlagSet).Var
                           at /usr/local/go1.24.0/src/flag/flag.go:1020:23
           3: 0x17461 - runtime.run$1
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:251:10
           4: 0x1503c - <goroutine wrapper>
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:250:2
           5:  0x7c9 - tinygo_launch
                           at /home/ayke/src/tinygo/tinygo/src/internal/task/task_asyncify_wasm.S:59
           6: 0x14eba - (*internal/task.Task).Resume
                           at /home/ayke/src/tinygo/tinygo/src/internal/task/task_asyncify.go:114:17              - runtime.scheduler
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:230:11              - runtime.run
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:255:11              - _start
                           at /home/ayke/src/tinygo/tinygo/src/runtime/runtime_wasmentry.go:20:5
    2: wasm trap: wasm `unreachable` instruction executed
panic: runtime error: nil pointer dereference
FAIL    compress/lzw    0.219s
```

deadprogram commented:
Perhaps the "important" part of the error is this?

```
# crypto/internal/sysrand
Error: /opt/hostedtoolcache/go/1.24.0/x64/src/crypto/internal/sysrand/rand_wasip1.go:14:17: undefined: syscall.RandomGet
```
That function is defined here:
https://go.googlesource.com/go/+/refs/tags/go1.24.0/src/syscall/fs_wasip1.go#946

How does `fs_wasip1.go` get included in the TinyGo build process again?

aykevl commented:
@deadprogram it's not, that's a separate issue. I just tracked it down to an interp bug (that I think I have a fix for).

aykevl commented:
See: https://github.com/tinygo-org/tinygo/pull/4742

aykevl committed:
runtime: add FIPS helper functions

Not entirely sure what they're for, but the Go runtime also stores this
information per goroutine so let's go with that.

aykevl committed:
syscall: add wasip1 RandomGet

This function is needed starting with Go 1.24.

aykevl commented:
Another scary error:

(EDIT: actually this is just missing panic/recover support on wasm)

```
~/src/tinygo/tinygo$ tinygo test -target=wasip1 crypto/des
Error: failed to run main module `/tmp/tinygo1749563909/main.wasmopt`

Caused by:
    0: failed to invoke command default
    1: error while executing at wasm backtrace:
           0: 0x232c5 - runtime.abort
                           at /home/ayke/src/tinygo/tinygo/src/runtime/runtime_tinygowasm.go:78:6              - runtime.panicOrGoexit
                           at /home/ayke/src/tinygo/tinygo/src/runtime/panic.go:81:7
           1: 0x3be4 - runtime._panic
                           at /home/ayke/src/tinygo/tinygo/src/runtime/panic.go:55:15
           2: 0x3f5c9 - (*crypto/des.desCipher).Encrypt              - (Go interface method)
                           at <Go interface method>
           3: 0x7148b - main!(crypto/cipher.Block).Encrypt$bound
           4: 0x7432a - crypto/internal/cryptotest.testCipher$6$1
                           at /usr/local/go1.24.0/src/crypto/internal/cryptotest/block.go:208:57
           5: 0x70125 - crypto/internal/cryptotest.mustPanic
                           at /usr/local/go1.24.0/src/crypto/internal/cryptotest/block.go:253:3
           6: 0x73bd4 - crypto/internal/cryptotest.testCipher$6
                           at /usr/local/go1.24.0/src/crypto/internal/cryptotest/block.go:208:12
           7: 0x20dc3 - main!testing.tRunner
           8: 0x55633 - (*testing.T).Run
           9: 0x7189f - crypto/internal/cryptotest.testCipher
                           at /usr/local/go1.24.0/src/crypto/internal/cryptotest/block.go:199:7
          10: 0x709be - crypto/internal/cryptotest.TestBlock$1
                           at /usr/local/go1.24.0/src/crypto/internal/cryptotest/block.go:31:13
          11: 0x20dc3 - main!testing.tRunner
          12: 0x55633 - (*testing.T).Run
          13: 0x706c1 - crypto/internal/cryptotest.TestBlock
                           at /usr/local/go1.24.0/src/crypto/internal/cryptotest/block.go:30:7
          14: 0x79ed1 - crypto/des_test.TestDESBlock$1
                           at /usr/local/go1.24.0/src/crypto/des/des_test.go:1513:23
          15: 0x20dc3 - main!testing.tRunner
          16: 0x55633 - (*testing.T).Run
          17: 0x79e18 - crypto/des_test.TestDESBlock
                           at /usr/local/go1.24.0/src/crypto/des/des_test.go:1512:7
          18: 0x20dc3 - main!testing.tRunner
          19: 0x55633 - (*testing.T).Run
          20: 0x20aa1 - testing.runTests$1
                           at /home/ayke/src/tinygo/tinygo/src/testing/testing.go:574:10
          21: 0x20dc3 - main!testing.tRunner
          22: 0x11ac8 - main!runtime.run$1
          23: 0xef02 - <goroutine wrapper>
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:250:2
          24:  0x6e6 - tinygo_launch
                           at /home/ayke/src/tinygo/tinygo/src/internal/task/task_asyncify_wasm.S:59
          25: 0xed82 - (*internal/task.Task).Resume
                           at /home/ayke/src/tinygo/tinygo/src/internal/task/task_asyncify.go:114:17              - runtime.scheduler
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:230:11              - runtime.run
                           at /home/ayke/src/tinygo/tinygo/src/runtime/scheduler_cooperative.go:255:11              - _start
                           at /home/ayke/src/tinygo/tinygo/src/runtime/runtime_wasmentry.go:20:5
    2: wasm trap: wasm `unreachable` instruction executed
panic: crypto/des: invalid buffer overlap
FAIL    crypto/des      0.191s
```

ydnar commented:
```
package go.bytecodealliance.org/wit
	imports bytes
	imports errors
	imports internal/reflectlite
	imports reflect
	imports iter
	imports runtime
	imports reflect: import cycle not allowed
```

Go 1.24 `reflect` now depends on on package `iter`.

TinyGo package `runtime` imports `reflect`, but big Go `runtime` does not.

deadprogram commented:
@aykevl I think you can rebase this against `dev` now to see where it stands....

aykevl commented:
Oh this is also going to be fun.

```
$ tinygo test -target=wasip2 crypto/des
FAIL    crypto/des      0.000s
# internal/syscall/unix
/usr/local/go1.24.0/src/internal/syscall/unix/at_sysnum_linux.go:12:35: undefined: syscall.SYS_READLINKAT
/usr/local/go1.24.0/src/internal/syscall/unix/at_sysnum_linux.go:13:35: undefined: syscall.SYS_MKDIRAT
/usr/local/go1.24.0/src/internal/syscall/unix/waitid_linux.go:17:82: undefined: syscall.Rusage
/usr/local/go1.24.0/src/internal/syscall/unix/waitid_linux.go:18:42: undefined: syscall.SYS_WAITID
```

aykevl commented:
Changed some package tests:

  * `crypto/des` now needs panic/recover to work so only running on non-wasm.
  * `crypto/rc4` now needs to be able to catch runtime panics (index out of range), which we haven't implemented on any platform. I've removed the test for now, we can re-enable it once we implement catching such runtime panics.

aykevl commented:
@ydnar how can I reproduce that issue? `-target=wasip2` seems to work fine for me.

aykevl commented:
Also, is there somebody who would like to help with wasip2 support? (I'll get there eventually, but it would speed things up).

```
$ tinygo test -target=wasip2 crypto/ecdsa
FAIL    crypto/ecdsa    0.000s
# internal/syscall/unix
/usr/local/go1.24.0/src/internal/syscall/unix/at_sysnum_linux.go:12:35: undefined: syscall.SYS_READLINKAT
/usr/local/go1.24.0/src/internal/syscall/unix/at_sysnum_linux.go:13:35: undefined: syscall.SYS_MKDIRAT
/usr/local/go1.24.0/src/internal/syscall/unix/waitid_linux.go:17:82: undefined: syscall.Rusage
/usr/local/go1.24.0/src/internal/syscall/unix/waitid_linux.go:18:42: undefined: syscall.SYS_WAITID
```

deadprogram commented:
@ydnar and @dgryski any chance you can look at the `wasip2` related errors in this PR?

ydnar commented:
There's a template package issue. It wants Seq/CanSeq.

I have a janky fix in progress. I can commit a draft PR if someone wants to take it across the finish line.

deadprogram commented:
> I can commit a draft PR if someone wants to take it across the finish line.

Yes, please, and thank you!

ydnar commented:
https://github.com/tinygo-org/tinygo/pull/4750

aykevl committed:
os: add File.Chdir support

We should really be using syscall.Fchdir here, but this is a fix to get
Go 1.24 working.

aykevl committed:
testing: add Chdir

This method was added in Go 1.24.

aykevl committed:
wasip2: add stubs to get internal/syscall/unix to work

This fixes lots of broken tests in stdlib packages in Go 1.24.

aykevl committed:
os: implement stub Chdir for non-OS systems

deadprogram commented:
https://github.com/tinygo-org/tinygo/actions/runs/13501641326/job/37722464866?pr=4732#step:7:57

It was going so well up to that point :crying_cat_face:

aykevl commented:
We should be getting close though!

aykevl commented:
So the Windows failure seems to be the only one left. And it doesn't have an error message :(

I'm afraid I'll have to reboot and fire up my Windows VM to debug this.

deadprogram commented:
@aykevl some clues here? https://go.dev/doc/go1.24#osuserpkgosuser

dgryski commented:
https://github.com/tinygo-org/tinygo/issues/3510 <-- tracking bug for recovering from runtime panics.

deadprogram commented:
One of the more recent commits here related? https://cs.opensource.google/go/go/+/release-branch.go1.24:src/os/user/user_windows_test.go;bpv=1

aykevl commented:
It's not a normal runtime panic (that would have printed the panic message at least). So #3510 is not going to help.

I was able to get it to run on Windows (Windows ARM anyway, that's all I have), and it's a segmentation fault. This is going to be fun to debug.

aykevl commented:
Update: actually it seems that I can't run _any_ tinygo binaries on Windows ARM. I think I'm running into a different bug specific to Windows ARM.

aykevl commented:
Getting a more useful error message when adding `GOARCH=amd64`:

<img width="788" alt="Screenshot 2025-02-24 at 20 37 47" src="https://github.com/user-attachments/assets/0f6d759f-48c4-4788-80db-5f14eaeb99b8" />

(Sorry about the screenshot, I couldn't get copy-paste to work).

I'm suspecting #4736

deadprogram commented:
@aykevl https://cs.opensource.google/go/go/+/release-branch.go1.24:src/os/user/user_windows_test.go;l=27-32;bpv=1

aykevl commented:
I think #4736 is broken, perhaps it would be best to revert this PR for now?
See: https://github.com/tinygo-org/tinygo/pull/4736#issuecomment-2679670226

deadprogram commented:
> I think https://github.com/tinygo-org/tinygo/pull/4736 is broken, perhaps it would be best to revert this PR for now?

Broken on Go 1.24 I assume you mean? It passed all of the CI tests at the time, which was part of why I merged it.

deadprogram commented:
@aykevl I tested this branch with PR #4736 reverted in PR #4755 and exactly same result.

deadprogram commented:
I can confirm that https://cs.opensource.google/go/go/+/5e82cba9bdf45d81da549477d172f6b5e23106d4 is what is triggering the Windows test failures. Looking deeper now into the root cause.

aykevl commented:
> Broken on Go 1.24 I assume you mean?

Broken in general (not just 1.24). I tried a trivial example, and it crashed. See my comments on that PR.

aykevl commented:
I removed the os/user test since it's not currently working (looks like they added the `t.Skip()` code in Go 1.24, and we currently don't seem to have a working implementation of that). I can look into adding `t.Skip()` support in the future, but for now it may be best to just skip this test entirely.

deadprogram reviewed (approved):
Wow, that was not easy! Great work all around on this one.

deadprogram commented:
@aykevl can you resolve merge conflict since I merged the revert into dev? I probably should have done it in the opposite order, oh well.

deadprogram commented:
@aykevl never mind I resolved merge conflict. Now just waiting for CI to run...

deadprogram commented:
Now merging! Thanks everyone.

deadprogram commented:
We will need to address https://github.com/tinygo-org/tinygo/issues/4759 before release.

aykevl commented:
@deadprogram see https://github.com/tinygo-org/tinygo/pull/4761

gandarez commented:
Since it's merged and blocked on https://github.com/tinygo-org/tinygo/issues/4759. Is it possible to build tinygo from source?

nvm I found it at BUILDING.md that's not referenced in the README.md.

aykevl commented:
@gandarez please take a look here: https://tinygo.org/docs/guides/build/
